### PR TITLE
Use lookup memory footprint in MSQ memory computations.

### PR DESCRIPTION
**Pull request by gianm**
Two main changes:

1) Add estimateHeapFootprint to LookupExtractor.

2) Use this in MSQ's IndexerWorkerContext when determining the total
   amount of available memory. It's taken off the top.

This prevents MSQ tasks from running out of memory when there are lookups defined in the cluster.

### CODE REVIEW

**Review by clintropolis — APPROVED** [review]
seems reasonable to me :metal:

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/IndexerWorkerContext.java [hunk 1]
```diff
@@ -276,6 +268,62 @@ public Bouncer processorBouncer()
...
+
+  /**
+   * Total estimated lookup footprint. Obtained by calling {@link LookupExtractor#estimateHeapFootprint()} on
+   * all available lookups.
+   */
+  private long computeTotalLookupFootprint()
+  {
+    // Subtract memory taken up by lookups. Correctness of this operation depends on lookups being loaded *before*
+    // we create this instance. Luckily, this is the typical mode of operation, since by default
+    // druid.lookup.enableLookupSyncOnStartup = true.
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/IndexerWorkerContext.java:301]: Should we add a comment about this in the memory tuning section of MSQ docs ?

Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/IndexerWorkerContext.java:301]: Also, should we just fetch this information from the coordinator in another PR? 
 Instead of calculating the size of each worker, the coordinator should be able to give us this size as part of lookup metadata ? wdyt ?

File: extensions-core/lookups-cached-single/src/main/java/org/apache/druid/server/lookup/cache/polling/OnHeapPollingCache.java [hunk 2]
```diff
@@ -83,6 +84,19 @@ public List<K> getKeys(final V value)
...
+  @SuppressWarnings("unchecked")
+  public long estimateHeapFootprint()
+  {
+    for (final Map.Entry<K, V> entry : immutableMap.entrySet()) {
+      if (!(entry.getKey() instanceof String) || !(entry.getValue() instanceof String)) {
+        return 0;
+      }
+    }
+
+    return MapLookupExtractor.estimateHeapFootprint((Map<String, String>) immutableMap);
```
Comment (cryptoe) [inline extensions-core/lookups-cached-single/src/main/java/org/apache/druid/server/lookup/cache/polling/OnHeapPollingCache.java:97]: IMHO, Instead of the monitor which runs every 10 mins, the place where we are updating the cache should do the running size calculation. Then this method just becomes a getter on top of that variable. 

We can do that as part of another PR as well.

File: extensions-core/lookups-cached-single/src/main/java/org/apache/druid/server/lookup/cache/polling/OnHeapPollingCache.java [hunk 3]
```diff
@@ -83,6 +84,19 @@ public List<K> getKeys(final V value)
     return listOfKeys;
   }
 
+  @Override
+  @SuppressWarnings("unchecked")
+  public long estimateHeapFootprint()
+  {
+    for (final Map.Entry<K, V> entry : immutableMap.entrySet()) {
+      if (!(entry.getKey() instanceof String) || !(entry.getValue() instanceof String)) {
```
[the remark this answers is not in the document]
Comment (gianm) [reply extensions-core/lookups-cached-single/src/main/java/org/apache/druid/server/lookup/cache/polling/OnHeapPollingCache.java:92]: Sounds reasonable. I modified the `MapLookupExtractor#estimateHeapFootprint` signature to `public static <K, V> long estimateHeapFootprint(@Nullable final Map<K, V> map)`, and note that it estimates zero footprint for nonstring stuff.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/IndexerWorkerContext.java (hunk 1 above)
Comment (gianm) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/IndexerWorkerContext.java:301]: Absolutely. I missed including this. It's updated in the latest version.

Comment (gianm) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/IndexerWorkerContext.java:301]: The Coordinator doesn't know how big the lookups are: it has the definitions, but nothing more. So I think we are doing about the best we can here.

File: extensions-core/lookups-cached-single/src/main/java/org/apache/druid/server/lookup/cache/polling/OnHeapPollingCache.java (hunk 2 above)
Comment (gianm) [reply extensions-core/lookups-cached-single/src/main/java/org/apache/druid/server/lookup/cache/polling/OnHeapPollingCache.java:97]: Makes sense to me. I'd rather that be a different PR, especially since I'm not really changing existing behavior in this PR. (Current code in master does the footprint estimation every time the monitor is run.)

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/IndexerWorkerContext.java [hunk 4]
```diff
@@ -250,8 +242,8 @@ public FrameContext frameContext(QueryDefinition queryDef, int stageNumber)
         indexIO,
         dataSegmentProvider,
         WorkerMemoryParameters.compute(
-            Runtime.getRuntime().maxMemory(),
-            numWorkersInJvm,
+            computeAvailableHeapMemory(),
```
[the remark this answers is not in the document]
Comment (gianm) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/IndexerWorkerContext.java:245]: It does need to be done on each worker, because the workers and the controller might have different amounts of memory available to them. (Imagine a job running through a rolling update that alters the amount of available memory.)

Good point that we should avoid needless recomputation of this. I updated things to compute it a single time per worker, as part of `IndexerWorkerContext#createProductionInstance`.

**Review by cryptoe — APPROVED** [review]
LGTM !!

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/IndexerWorkerContext.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (cryptoe) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/IndexerWorkerContext.java]: I meant doing only the lookup sizing calculation on the controller but I see why it can be misleading :). 
Thanks for the fixes.
